dlt/destinations/impl/clickhouse/clickhouse.py

```diff
@@ -0,0 +1,402 @@
+                secret_access_key = staging_credentials.aws_secret_access_key
+            elif isinstance(staging_credentials, GcpCredentials):
+                access_key_id = dlt.config["destination.filesystem.credentials.gcp_access_key_id"]
+                secret_access_key = dlt.config[
+                    "destination.filesystem.credentials.gcp_secret_access_key"
+                ]
+
+            auth = "NOSIGN"
+            if access_key_id and secret_access_key:
+                auth = f"'{access_key_id}','{secret_access_key}'"
+
+            table_function = (
+                f"s3('{bucket_http_url}',{auth},'{clickhouse_format}','auto','{compression}')"
+            )
+
+        elif bucket_scheme in ("az", "abfs"):
+            if not isinstance(staging_credentials, AzureCredentialsWithoutDefaults):
+                raise LoadJobTerminalException(
+                    file_path,
+                    "Unsigned Azure Blob Storage access from ClickHouse isn't supported as yet.",
+                )
+
+            # Authenticated access.
+            account_name = staging_credentials.azure_storage_account_name
+            storage_account_url = (
+                f"https://{staging_credentials.azure_storage_account_name}.blob.core.windows.net"
+            )
+            account_key = staging_credentials.azure_storage_account_key
+
+            # build table func
+            table_function = f"azureBlobStorage('{storage_account_url}','{bucket_url.netloc}','{bucket_url.path}','{account_name}','{account_key}','{clickhouse_format}','{compression}')"
+        else:
+            raise LoadJobTerminalException(
+                file_path,
+                f"ClickHouse loader does not support '{bucket_scheme}' filesystem.",
+            )
+
+        statement = f"INSERT INTO {qualified_table_name} SELECT * FROM {table_function}"
+        with client.begin_transaction():
+            client.execute_sql(statement)
+
+    def state(self) -> TLoadJobState:
+        return "completed"
+
+    def exception(self) -> str:
+        raise NotImplementedError()
+
+
+class ClickHouseMergeJob(SqlMergeJob):
+    @classmethod
+    def _to_temp_table(cls, select_sql: str, temp_table_name: str) -> str:
+        return f"CREATE TABLE {temp_table_name} ENGINE = Memory AS {select_sql};"
+
+    @classmethod
+    def gen_key_table_clauses(
+        cls,
+        root_table_name: str,
+        staging_root_table_name: str,
+        key_clauses: Sequence[str],
+        for_delete: bool,
+    ) -> List[str]:
+        join_conditions = " AND ".join([c.format(d="d", s="s") for c in key_clauses])
+        return [
+            f"FROM {root_table_name} AS d JOIN {staging_root_table_name} AS s ON {join_conditions}"
+        ]
+
+    @classmethod
+    def gen_update_table_prefix(cls, table_name: str) -> str:
+        return f"ALTER TABLE {table_name} UPDATE"
+
+    @classmethod
+    def requires_temp_table_for_delete(cls) -> bool:
```

Comment:
this will make the merge job use the merge strategy for table with child tables which does the delete also on the root table with a temp table. I could not get clickhouse to do a delete in one sql call when selecting multiple keys, so this was the easiest way. The only drawback is, that you will need to have a unique column defined.

Reply:
I can't find code generating MERGE sql. but if you do it then it is completely fine

Reply:
no, what I meant to say, it will fallback to the code that is used to to a merge for a tablechain with more than one table. there are no MERGE statements.